### DISCUSSION TITLE
Questions about training self-generated signal images.

### DISCUSSION THREAD

**Discussion by Minxiangliu** [ultralytics/yolov5#9509] [Q&A]
Many thanks for this great tool that can speed up the development process.
In my research, my data is some raw signal data, I can use `python`, `matplotlib`, `opencv`, etc. Using these tools I can draw the signal shape for people to observe. Now I want to use YOLO to find some interesting pattern. I would like to ask what are the suggestions here what kind of images should I draw so that the model can learn better, given that they are all correctly labeled?
Color, line thickness, image size?
I am currently drawing a color image, the size is `512x256`, and then the size is adjusted to `640` when the model is trained.

In fact, I have tried `other models`, `preset augmentation scripts`, and even `parameter evolution`, and the number of labels has continued to increase, but recently encountered a bottleneck. No matter how the model is trained, the accuracy result is about 80%, and there is no way to approach 90%. I know my training is still overfitting and I know it's not close to the amount of training you suggest, but I'm wondering if increasing the amount of data is the only way?

Thanks in advance!

Below is what I drew:
![U121167900_1](https://user-images.githubusercontent.com/30144428/191199400-dc02de87-9cb1-4718-a73d-864c813b7cf6.png)

![U121167900_2](https://user-images.githubusercontent.com/30144428/191199424-e9856203-957f-4a72-8818-b70965df0aa9.png)

possible labels:
![image](https://user-images.githubusercontent.com/30144428/191199773-a6d34d20-7b82-43ee-964b-70489ba0a279.png)

![image](https://user-images.githubusercontent.com/30144428/191199920-7fdebb77-f11c-4a9c-8c9e-a3ed5b25d90e.png)

My Data:
Image number: 3013
Unlabeled negative image number: 176
label 1 number: 3202
label 2 number: 3213

My Train:
Randomly split positive and negative image -> train: 0.9, test: 0.1
Training command:
`python train.py --img 640 --batch 64 --epochs 600 --data datasets.yaml --weights yolov5n.pt --workers 4 --hyp hyp.scratch-high.yaml`

Results:
![image](https://user-images.githubusercontent.com/30144428/191209454-f090abc2-5640-47fe-8de4-54f8ca3ad3d9.png)

![image](https://user-images.githubusercontent.com/30144428/191209788-f00e644d-522e-4b19-ad2d-917476313bbc.png)

![image](https://user-images.githubusercontent.com/30144428/191210009-e8e6e9f6-907a-495a-a791-ea29960a3756.png)

**Comment by glenn-jocher** [comment]
@Minxiangliu 👋 Hello! Thanks for asking about improving YOLOv5 🚀 training results. 

Most of the time good results can be obtained with no changes to the models or training settings, **provided your dataset is sufficiently large and well labelled**. If at first you don't get good results, there are steps you might be able to take to improve, but we always recommend users **first train with all default settings** before considering any changes. This helps establish a performance baseline and spot areas for improvement.

If you have questions about your training results **we recommend you provide the maximum amount of information possible** if you expect a helpful response, including results plots (train losses, val losses, P, R, mAP), PR curve, confusion matrix, training mosaics, test results and dataset statistics images such as labels.png. All of these are located in your `project/name` directory, typically `yolov5/runs/train/exp`.

We've put together a full guide for users looking to get the best results on their YOLOv5 trainings below.

## Dataset

- **Images per class.** ≥ 1500 images per class recommended
- **Instances per class.** ≥ 10000 instances (labeled objects) per class recommended
- **Image variety.** Must be representative of deployed environment. For real-world use cases we recommend images from different times of day, different seasons, different weather, different lighting, different angles, different sources (scraped online, collected locally, different cameras) etc.
- **Label consistency.** All instances of all classes in all images must be labelled. Partial labelling will not work.
- **Label accuracy.** Labels must closely enclose each object. No space should exist between an object and it's bounding box. No objects should be missing a label.
- **Label verification.** View `train_batch*.jpg` on train start to verify your labels appear correct, i.e. see [example](https://github.com/ultralytics/yolov5/wiki/Train-Custom-Data#local-logging) mosaic.
- **Background images.** Background images are images with no objects that are added to a dataset to reduce False Positives (FP). We recommend about 0-10% background images to help reduce FPs (COCO has 1000 background images for reference, 1% of the total). No labels are required for background images.

<a href="https://arxiv.org/abs/1405.0312"><img width="800" src="https://user-images.githubusercontent.com/26833433/109398377-82b0ac00-78f1-11eb-9c76-cc7820669d0d.png" alt="COCO Analysis"></a>


## Model Selection

Larger models like YOLOv5x and [YOLOv5x6](https://github.com/ultralytics/yolov5/releases/tag/v5.0) will produce better results in nearly all cases, but have more parameters, require more CUDA memory to train, and are slower to run. For **mobile** deployments we recommend YOLOv5s/m, for **cloud** deployments we recommend YOLOv5l/x. See our README [table](https://github.com/ultralytics/yolov5#pretrained-checkpoints) for a full comparison of all models. 

<p align="center"><img width="700" alt="YOLOv5 Models" src="https://github.com/ultralytics/yolov5/releases/download/v1.0/model_comparison.png"></p>

- **Start from Pretrained weights.** Recommended for small to medium sized datasets (i.e. [VOC](https://github.com/ultralytics/yolov5/blob/master/data/VOC.yaml), [VisDrone](https://github.com/ultralytics/yolov5/blob/master/data/VisDrone.yaml), [GlobalWheat](https://github.com/ultralytics/yolov5/blob/master/data/GlobalWheat2020.yaml)). Pass the name of the model to the `--weights` argument. Models download automatically from the [latest YOLOv5 release](https://github.com/ultralytics/yolov5/releases).
```bash
python train.py --data custom.yaml --weights yolov5s.pt
                                             yolov5m.pt
                                             yolov5l.pt
                                             yolov5x.pt
                                             custom_pretrained.pt
```
- **Start from Scratch.** Recommended for large datasets (i.e. [COCO](https://github.com/ultralytics/yolov5/blob/master/data/coco.yaml), [Objects365](https://github.com/ultralytics/yolov5/blob/master/data/Objects365.yaml), [OIv6](https://storage.googleapis.com/openimages/web/index.html)). Pass the model architecture yaml you are interested in, along with an empty `--weights ''` argument:
```bash
python train.py --data custom.yaml --weights '' --cfg yolov5s.yaml
                                                      yolov5m.yaml
                                                      yolov5l.yaml
                                                      yolov5x.yaml
```


## Training Settings

Before modifying anything, **first train with default settings to establish a performance baseline**. A full list of train.py settings can be found in the [train.py](https://github.com/ultralytics/yolov5/blob/master/train.py) argparser.

- **Epochs.** Start with 300 epochs. If this overfits early then you can reduce epochs. If overfitting does not occur after 300 epochs, train longer, i.e. 600, 1200 etc epochs.
- **Image size.** COCO trains at native resolution of `--img 640`, though due to the high amount of small objects in the dataset it can benefit from training at higher resolutions such as `--img 1280`. If there are many small objects then custom datasets will benefit from training at native or higher resolution. Best inference results are obtained at the same `--img` as the training was run at, i.e. if you train at `--img 1280` you should also test and detect at `--img 1280`.
- **Batch size.** Use the largest `--batch-size` that your hardware allows for. Small batch sizes produce poor batchnorm statistics and should be avoided.
- **Hyperparameters.** Default hyperparameters are in [hyp.scratch-low.yaml](https://github.com/ultralytics/yolov5/blob/master/data/hyps/hyp.scratch-low.yaml). We recommend you train with default hyperparameters first before thinking of modifying any. In general, increasing augmentation hyperparameters will reduce and delay overfitting, allowing for longer trainings and higher final mAP. Reduction in loss component gain hyperparameters like `hyp['obj']` will help reduce overfitting in those specific loss components. For an automated method of optimizing these hyperparameters, see our [Hyperparameter Evolution Tutorial](https://github.com/ultralytics/yolov5/issues/607).

## Further Reading

If you'd like to know more a good place to start is Karpathy's 'Recipe for Training Neural Networks', which has great ideas for training that apply broadly across all ML domains: http://karpathy.github.io/2019/04/25/recipe/

Good luck 🍀 and let us know if you have any other questions!
